src/caffeine/caffeine.c

```diff
@@ -157,9 +137,61 @@ void c_co_broadcast(CFI_cdesc_t * a_desc, int source_image, int* stat, int num_e
      size_t c_sizeof_a = a_desc->elem_len;
      int nbytes = num_elements * c_sizeof_a;
   
-     gex_Event_t ev
-       = gex_Coll_BroadcastNB(myteam, source_image-1, c_loc_a, c_loc_a, nbytes, 0);
-     gex_Event_Wait(ev);  
+  int data_type = a_desc->type;
+
+  gex_Event_t ev
+    = gex_Coll_BroadcastNB(myteam, source_image-1, c_loc_a, c_loc_a, nbytes, 0);
+  gex_Event_Wait(ev);  
+
+  if (stat != NULL) *stat = 0;
+}
+
+void c_co_sum(CFI_cdesc_t* a_desc, int result_image, int* stat, char* errmsg, size_t num_elements)
+{
+  CFI_type_t a_type = a_desc->type;
+  size_t c_sizeof_a;
+
+  switch (a_type)
+  {
+    case CFI_type_int32_t:        a_type = GEX_DT_I32; break;
+    case CFI_type_int64_t:        a_type = GEX_DT_I64; break;
+    case CFI_type_float:          a_type = GEX_DT_FLT; break;
+    case CFI_type_double:         a_type = GEX_DT_DBL; break;
+#if __GNUC__ >= 12
+    case CFI_type_float_Complex:  a_type = GEX_DT_FLT; num_elements *= 2; break;
+    case CFI_type_double_Complex: a_type = GEX_DT_DBL; num_elements *= 2; break;
+#else
+    case 2052:                    a_type = GEX_DT_FLT; num_elements *= 2; break;
+    case 4100:                    a_type = GEX_DT_DBL; num_elements *= 2; break;
+#endif
```

Comment:
Ick to these constants.

I suggest #defining macros called CAF_CFI_type_{float,double}_Complex at the top of this file and use those here.